src/trace_writer.ml

```diff
@@ -821,24 +846,47 @@ let write_event (T t) event =
       | Some ip -> is_kernel_address ip
     in
     end_of_thread t thread_info ~time ~is_kernel_address
-  | Ok (Power { thread = _; time = _; freq }) ->
+  | Ok
+      { Event.Ok.thread = _ (* Already used this to look up thread info. *)
+      ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
+      ; data = Power { freq }
+      } ->
     write_counter
       t
       ~thread
       ~name:"CPU"
       ~time
       ~args:Tracing.Trace.Arg.[ "freq (MHz)", Int freq ]
-  | Ok (Trace event) ->
-    let { Event.Ok.Trace.thread = _ (* Already used this to look up thread info. *)
-        ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
-        ; kind
-        ; trace_state_change
-        ; src
-        ; dst
-        }
-      =
-      event
+  | Ok
+      { Event.Ok.thread = _ (* Already used this to look up thread info. *)
+      ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
+      ; data = Sample { callstack }
+      } ->
+    let how_many_ret =
+      Stack.length thread_info.callstack.stack
+      - Callstack.how_many_match thread_info.callstack callstack
+    in
+    List.init how_many_ret ~f:Fn.id |> List.iter ~f:(fun _ -> ret t thread_info ~time);
+    let how_many_call =
+      List.length callstack - Stack.length thread_info.callstack.stack
+    in
+    let calls = List.drop callstack (Stack.length thread_info.callstack.stack) in
+    let (_ : Event.Location.t list) =
+      List.init how_many_call ~f:Fn.id
+      |> List.fold ~init:calls ~f:(fun calls _ ->
+             match calls with
+             | head :: tail ->
+               call t thread_info ~time ~location:head;
+               tail
+             | [] ->
+               raise_s [%message "BUG: Unexpected behavior while parsing callstack."])
```

Comment:
I think we should print some more information as part of this message, to make inevitable bug reports easier to track down.

Reply:
This raise_s no longer exists and the loop is just an iter (ended up awkward because it wasn't looping the entire list when written).